client/src/embedded_support.ts

```diff
@@ -0,0 +1,88 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+import * as ts from 'typescript';
+
+/** Determines if the position is inside an inline template. */
+export function isInsideInlineTemplateRegion(documentText: string, offset: number): boolean {
+  return isPropertyAssignmentToStringOrStringInArray(documentText, offset, ['template']);
+}
+
+/** Determines if the position is inside an inline template, templateUrl, or string in styleUrls. */
+export function isInsideAngularContext(documentText: string, offset: number): boolean {
+  return isPropertyAssignmentToStringOrStringInArray(
+      documentText, offset, ['template', 'templateUrl', 'styleUrls']);
+}
+
+/**
+ * Basic scanner to determine if we're inside a string of a property with one of the given names.
+ *
+ * This scanner is not currently robust or perfect but provides us with an accurate answer _most_ of
+ * the time.
+ *
+ * False positives are OK here. Though this will give some false positives for determining if a
+ * position is within an Angular context, i.e. an object like `{template: ''}` that is not inside an
+ * `@Component` or `{styleUrls: [someFunction('stringL¦iteral')]}`, the @angular/language-service
+ * will always give us the correct answer. This helper gives us a quick win for optimizing the
+ * number of requests we send to the server.
+ */
+function isPropertyAssignmentToStringOrStringInArray(
+    documentText: string, offset: number, propertyAssignmentNames: string[]): boolean {
+  const scanner = ts.createScanner(ts.ScriptTarget.ESNext, true /* skipTrivia */);
+  scanner.setText(documentText);
+
+  let token: ts.SyntaxKind = scanner.scan();
+  let lastToken: ts.SyntaxKind|undefined;
+  let lastTokenText: string|undefined;
+  let unclosedBraces = 0;
+  let unclosedBrackets = 0;
+  let propertyAssignmentContext = false;
+  while (token !== ts.SyntaxKind.EndOfFileToken && scanner.getStartPos() < offset) {
+    if (lastToken === ts.SyntaxKind.Identifier &&
+        propertyAssignmentNames.some(name => name === lastTokenText) &&
+        token === ts.SyntaxKind.ColonToken) {
+      propertyAssignmentContext = true;
+      token = scanner.scan();
+      continue;
+    }
+    if (unclosedBraces === 0 && unclosedBrackets === 0 && isPropertyAssignmentTerminator(token)) {
+      propertyAssignmentContext = false;
+    }
+
+    if (token === ts.SyntaxKind.OpenBracketToken) {
+      unclosedBrackets++;
+    }
+    if (token === ts.SyntaxKind.OpenBraceToken) {
+      unclosedBraces++;
+    }
+    if (token === ts.SyntaxKind.CloseBracketToken) {
+      unclosedBrackets--;
+    }
+    if (token === ts.SyntaxKind.CloseBraceToken) {
+      unclosedBraces--;
+    }
+
+    const isStringToken = token === ts.SyntaxKind.StringLiteral ||
+        token === ts.SyntaxKind.NoSubstitutionTemplateLiteral;
+    const isCursorInToken = scanner.getStartPos() <= offset &&
+        scanner.getStartPos() + scanner.getTokenText().length >= offset;
```

Comment:
Should the `offset` be strictly less than `scanner.getStartPos() + scanner.getTokenText().length`?

Reply:
Added test with comment about why we include that position.